src/util/compat.hpp

```diff
@@ -34,4 +34,27 @@ inline void compiler_barrier_if_xlc_leq(unsigned ver) {
 template <typename X>
 inline constexpr bool isinf(X x) { return std::isinf(x); }
 
+// Work around a bad inlining-related optimization with icpc 16.0.3 and -xMIC-AVX512,
+// by forcing a computation.
+
+template <typename X>
+inline void sink(const X& x) {
```

Comment:
Actually, I'm not sure about the reliability of just copying one byte from `x`; while it works in this particular instance, the bug presented itself as in fact getting one field of the pair correct and not the other.

But I will replace the `char` array with a single `char`.